src/lamp_py/bus_performance_manager/vehicle_events.py

```diff
@@ -152,3 +155,115 @@ def get_service_date_from_filename(tm_filename: str) -> Optional[date]:
     )
 
     return grouped_files.to_dicts()
+
+
+def bus_routes_for_service_date(service_date: date) -> List[str]:
+    """get a list of bus route ids for a given service date"""
+    routes_file = get_gtfs_parquet_file(
+        year=service_date.year, filename="routes.parquet"
+    ).get_s3_path()
+
+    # generate an integer date for the service date
+    target_date = int(service_date.strftime("%Y%m%d"))
+
+    bus_routes = (
+        pl.scan_parquet(routes_file)
+        .select(["route_id", "route_type", "gtfs_active_date", "gtfs_end_date"])
+        .filter(
+            (pl.col("gtfs_active_date") < target_date)
+            & (pl.col("gtfs_end_date") > target_date)
+            & (pl.col("route_type") == 3)
+        )
+        .collect()
+    )
+
+    return bus_routes["route_id"].unique().to_list()
+
+
+def generate_gtfs_rt_events(
+    service_date: date, gtfs_rt_files: List[str]
+) -> pl.DataFrame:
+    """
+    generate a polars dataframe for bus vehicle events from gtfs realtime
+    vehicle position files for a given service date
+    """
+    bus_routes = bus_routes_for_service_date(service_date)
+
+    # build a dataframe of every gtfs record
+    # * scan all of the gtfs realtime files
+    # * only pull out bus records and records for the service date
+    # * rename / convert columns as appropriate
+    # * sort by vehicle id and timestamp
+    # * keep only the first record for a given trip / stop / status
+    # * sort again by vehicle id and timestamp (the unique call seems to reorder)
+    vehicle_positions = (
+        pl.scan_parquet(gtfs_rt_files)
+        .filter(
+            (pl.col("vehicle.trip.route_id").is_in(bus_routes))
+            & (
+                pl.col("vehicle.trip.start_date")
+                == service_date.strftime("%Y%m%d")
+            )
+        )
+        .select(
+            pl.col("vehicle.trip.route_id").cast(pl.String).alias("route_id"),
+            pl.col("vehicle.trip.trip_id").cast(pl.String).alias("trip_id"),
+            pl.col("vehicle.stop_id").cast(pl.String).alias("stop_id"),
+            pl.col("vehicle.trip.direction_id").alias("direction_id"),
+            pl.col("vehicle.trip.start_time").alias("start_time"),
+            pl.col("vehicle.trip.start_date").alias("service_date"),
+            pl.col("vehicle.vehicle.id").cast(pl.String).alias("vehicle_id"),
+            pl.col("vehicle.current_status").alias("current_status"),
+            pl.from_epoch("vehicle.timestamp")
+            .dt.convert_time_zone("EST5EDT")
+            .alias("vehicle_timestamp"),
+        )
+        .unique()
+        .sort(["vehicle_id", "vehicle_timestamp"])
+        .unique(
+            subset=[
+                "route_id",
+                "trip_id",
+                "stop_id",
+                "direction_id",
+                "vehicle_id",
+                "current_status",
+            ],
+            keep="first",
+        )
+        .sort(["vehicle_id", "vehicle_timestamp"])
```

Comment:
Probably don't need to do this `sort` here, as it will likely get un-sorted when joined to transit master data and then re-sorted for another purpose.